### ISSUE TITLE
"vagrant box remove" should remove image from libvirt pool

### ISSUE THREAD

**Issue by lelutin** [vagrant-libvirt/vagrant-libvirt#85]
When asking vagrant to remove a libvirt-specific base box, the corresponding image file should be removed from the libvirt pool.

That would avoid cluttering the disk with old image files, and would avoid user confusion when discovering that the box is still present somewhere even though vagrant said it was "removed".

**Comment by dennisklein** [comment]
> I could run `vagrant box add`, then run `vagrant up` with two Vagrantfiles that upload the same box to `qemu:///system` and `qemu+ssh://example.com/system`.

This usecase seems a bit out of scope for what vagrant was intended to be. @sciurus I would be interested in a more detailed example to understand when you need this.

> Related to this, I am intrigued by the possibility of no longer storing boxes in `~/.vagrant.d/boxes/` at all but instead having vagrant box always operate directly on a libvirt storage pool.

I like this approach. As purpleidea stated, saving the image twice seems like an inefficient way to go. Most users will not use multiple and/or remote hypervisors.

**Comment by aspiers** [comment]
> > Related to this, I am intrigued by the possibility of no longer storing boxes in ~/.vagrant.d/boxes/ at all but instead having vagrant box always operate directly on a libvirt storage pool.
> I like this approach.  As purpleidea stated, saving the image twice seems like an inefficient way to go.

True, it's a nice thought, but ...

> Most users will not use multiple and/or remote hypervisors.

In other words, *some* users *will* use multiple hypervisors, in which case at least one is remote.  And the problem is that Vagrant doesn't support a model where boxes can be stored in different locations.  You'd need to extend commands like `vagrant box list` so that they would support a `--hypervisor` option or similar, but only for the `libvirt` provider.  And I'm not sure if that's even possible or something Mitchell would want to entertain, because it's a fundamentally different model.  Currently box storage is flat, and local.

**Comment by sciurus** [comment]
@dennisklein [this](https://github.com/pradels/vagrant-libvirt/issues/85#issuecomment-56456372) isn't about something I need. In that comment I was just trying to explain what the current design allows and how it makes pool cleanup difficult.

If we were going by what I need we would rip out support for remote libvirt altogether. :smile:

**Comment by dustymabe** [comment]
I think what we need here is a two pronged approach:

1. store information about which libvirt URIs have been uploaded to and query the user (y/n) if they'd like to remove them from libvirt on `vagrant box remove`.
1. upload the box to the libvirt storage pool based on hash+name
    - since versioning isn't supported for boxes that don't get pulled from atlas #294 doesn't help us in all scenarios (see this mail list [thread](https://www.redhat.com/archives/container-tools/2016-June/msg00000.html))
    - compute a hash (one time upon vagrant box add) and name the volume in the libvirt storage pool accordingly
        - so if a box was originally going to be named 'cdkv2_vagrant_box_image_0.img' in the pool it could now be named cdkv2_vagrant_box_image_0_fb28e33f098abaf356940fdb3aaae417.img

Does this seem reasonable?

**Comment by LalatenduMohanty** [comment]
> Store information about which libvirt URIs have been uploaded to and query the user (y/n) if they'd like to remove them from libvirt on vagrant box remove.

I would be happy if we can just fix this issue as our first attempt. 

I can see some interesting discussion in the issue, but we do not have a solution yet. I think we have increased the expected  fix's scope more than what can be done by a contributor in a reasonable time frame.

**Comment by dustymabe** [comment]
> I would be happy if we can just fix this as our first attempt.

@LalatenduMohanty I think we need both. It's possible that the user will either answer "no" to removing from the libvirt provider or that a libvirt host is down during the time of the `vagrant box remove`. Identifying by content hash would mitigate this  corner case as well.

**Comment by jasonbrooks** [comment]
@dustymabe @LalatenduMohanty Check out sections 4 and 5 here: https://github.com/hollodotme/Helpers/blob/master/Tutorials/vagrant/self-hosted-vagrant-boxes-with-versioning.md It should be possible host w/ versioning. I did I quick test locally, it does seem to work...

**Comment by dustymabe** [comment]
@jasonbrooks interesting. I guess I assumed that it could not be done because of this error message:

```
$ vagrant box add --name cdk --box-version 2 ./rhel-cdk-kubernetes-7.2-23.x86_64.vagrant-libvirt.box
You specified a box version constraint with a direct box file
path. Box version constraints only work with boxes from Vagrant
Cloud or a custom box host. Please remove the version constraint
and try again.
```

Either way I still think there is something that can be done for making this a better user experience.

**Comment by jasonbrooks** [comment]
@dustymabe Yep, it just turns out that it's pretty easy to do a custom box host. It's a short json file, and if your box url location points to that json file, and the json file points to the various box versions, it works.

**Comment by daks** [comment]
Hello,
I think this feature would be really nice to have, but in the mean time, is there anyway to do it in a script?
I'm using vagrant-libvirt in some Jenkins jobs with custom boxes. When we upgrade this boxes, I can "vagrant box update" the Vagrant box, but it seems to not update the image on the libvirt pool, so I thought a solution would be to:
- remove Vagrant box
- remove image in libvirt storage
- update/install vagrant box
My problem is: how can I find the name of the image from the Vagrant name: is there any way to ask Vagrant this information?

**Comment by dustymabe** [comment]
hey @daks. If you query libvirt for the volumes in the default storage pool you should be able to identify the one to remove from the list. So `virsh vol-list --pool default` and then delete the volume with for example: `virsh vol-delete --pool default f25-20161007_vagrant_box_image_0.img`

**Comment by daks** [comment]
I saw this virsh functionality. What I want is, knowing my box name (e.g foo/bar), how to know what is the image name.
It seems that I can grep foo-VAGRANTSLASH-bar, but I asked if there was a better/more reliable way.

**Comment by dustymabe** [comment]
> But I asked to know is there is a more reliable way to find this filename.

not that I know of

**Comment by daks** [comment]
As removal of old image in libvirt storage seems not feasible, but now that box upgrade warning seems OK (see #713 ), I think most functionality is present, and this issue may be closed.

**Comment by axelsimon** [comment]
Hi everyone, sorry to revive an old issue, but i'm trying to wrap my mind around where Vagrant and libvirt put together store both images and VMs.

I have the following:
```
$ ls -lih .local/share/libvirt/images/
total 18G
8003751 -rw-------. 1 axsimon axsimon 2.7G Oct 14 17:59 ansible-keylime-tpm-emulator_keylime1.img
8002965 -rw-------. 1 axsimon axsimon  12G Oct 14 17:54 debian_default.img
8002962 -rwxr--r--. 1 axsimon axsimon 1.1G Jun 24 19:01 debian-VAGRANTSLASH-buster64_vagrant_box_image_10.4.0.img
8001134 -rwxr--r--. 1 axsimon axsimon 2.1G Jul  8 11:12 fedora-VAGRANTSLASH-31-cloud-base_vagrant_box_image_31.20191023.0.img

```
and
```
$ ls -lih .vagrant.d/boxes/**/*/libvirt/
.vagrant.d/boxes/debian-VAGRANTSLASH-buster64/10.4.0/libvirt/:
total 1.1G
8433562 -rw-r--r--. 1 axsimon axsimon 1.1G Jun 24 19:01 box.img
(…)

.vagrant.d/boxes/fedora-VAGRANTSLASH-31-cloud-base/31.20191023.0/libvirt/:
total 2.1G
8240060 -rwxr-xr-x. 1 axsimon axsimon 2.1G Jul  8 11:12 box.img
(…)
```

That is, i have twice the same img files in both locations. I checked, there is no difference between `fedora-VAGRANTSLASH-31-cloud-base_vagrant_box_image_31.20191023.0.img` and the `box.img` in `.vagrant.d/boxes/fedora-VAGRANTSLASH-31-cloud-base/31.20191023.0/libvirt/`, for instance. 

As you can see, they have different inodes too, so they are not hardlinks to the same file.

So, which ones does vagrant _actually_ use?

I'm thinking it's the ones in `.vagrant.d`, as when destroyed past box, its image was still in `.local/share/libvirt/images/`.

Any confirmation or help appreciated!

**Comment by daks** [comment]
Vagrant, by default and for any provider, stores images in `.vagrant.d`. When using it with libvirt it uploads it to the libvirt pool storage, which is local (`.local/share/libvirt/images` in your example) but could also be remote if using this functionality.

When you `vagrant box remove` it removes it from `.vagrant.d` directory but not your libvirt pool storage, which is the reason this issue was opened.

**Comment by axelsimon** [comment]
Thanks, and thanks for clarifying the issue at hand in this thread, @daks.

Not to piggyback too much on the thread, but is it considered normal that each image therefore gets stored twice on disk? That seems quite obviously wasteful. If it's going to be the exact same file, a link of some sort would make sense. I realise i'm essentially popping up in this years-old thread, so sorry if this has all been said before.

**Comment by mgedmin** [comment]
Until this is fixed, I've written a Python 3 script that matches libvirt image names against vagrant box names and deletes them if you ask.  In case this is helpful to anyone, here it is: https://gist.github.com/mgedmin/3ca0e89a9d8cd452c6e5bf8babe055dc

Usage:

    python3 prune_libvirt_images.py --verbose                           # prints everything
    python3 prune_libvirt_images.py --matching VAGRANTSLASH             # prints unused ones
    python3 prune_libvirt_images.py --matching VAGRANTSLASH --delete    # runs virsh vol-delete for each of them

**Comment by electrofelix** [comment]
@mgedmin thanks this will help people until something can be implemented that is a bit more intelligent.

Essentially I think that for this to work as requested the following will be needed:
* register a before hook on the vagrant box remove action
   * for the box being removed for each machine in the global index
      * check for machines using the relevant box
      * use the connection parameters from it's associated Vagrantfile (if it exists and is valid)
         * query libvirt for the relevant storage pools
         * query libvirt as to whether the expected image exists
         * query libvirt if the image is in use or not as a backing image <-- not sure if this is possible but presume it should be
         * if not, remove the backing image
* on vagrant destroy perform the same checks to see if libvirt reports any machine using the backing image and remove if not
   * provide a config option to toggle this behaviour
* register a custom command to perform clean up as it's always likely between the option and remote instances for local state to become invalid or for a new version of a box to be loaded in without the old version being explicitly removed.

**Comment by chaychoong** [comment]
Just encountered this issue. This could cause https://github.com/hashicorp/vagrant/issues/5186 if you attempt to remove an old base box, repackage and add a new one, and then attempt to run an application based off the base box.

**Comment by electrofelix** [comment]
@chaychoong PR #1382 should be preventing that from occurring by ensuring either the box version or the image file mtime is used to avoid such collisions since version 0.6.3. if you repackage an old base box there should be no version copied, and it'll switch to using the image file mtime. Only if you repackage a box and explicitly reuse the same version should it be possible to bump into such behaviour.

Relying on removal here wouldn't be the solution to such a problem, because if the image was in use by another machine it would not be possible to delete the base image anyway.

So if you are using a plugin version >= 0.6.3, please create a separate issue with a debug log as it means the collision avoidance would need more work. Ideally, please confirm with the latest release 0.8.1.

**Comment by chaychoong** [comment]
@electrofelix I don't have any versioning configs defined explicitly anywhere in my Vagrantfile, but it appears that the boxes I packaged is tagged with version `0`.

**Comment by electrofelix** [comment]
@chaychoong that suggests you are running an older version of the plugin and updating to 0.6.3 or newer will resolve your issue.

You should be able to see which version of the plugin is installed by using:
```
vagrant plugin list
```

**Comment by slwr4thales** [comment]
Issue still appear, backingStore file will not removed with

❯ uname -a
`Linux <value host name> 5.10.0-10-amd64 #1 SMP Debian 5.10.84-1 (2021-12-08) x86_64 GNU/Linux`

❯ cat /etc/os-release
```
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

❯ cat /etc/debian_version
`11.2`

❯ dpkg-query -l | grep -e libvirt -e vagrant
```
ii  gir1.2-libvirt-glib-1.0:amd64                               3.0.0-1                                amd64        GObject introspection files for the libvirt-glib library
ii  libsys-virt-perl                                            7.0.0-1                                amd64        Perl module providing an extension for the libvirt library
ii  libvirt-clients                                             7.0.0-3                                amd64        Programs for the libvirt library
ii  libvirt-daemon                                              7.0.0-3                                amd64        Virtualization daemon
ii  libvirt-daemon-config-network                               7.0.0-3                                all          Libvirt daemon configuration files (default network)
ii  libvirt-daemon-config-nwfilter                              7.0.0-3                                all          Libvirt daemon configuration files (default network filters)
ii  libvirt-daemon-driver-lxc                                   7.0.0-3                                amd64        Virtualization daemon LXC connection driver
ii  libvirt-daemon-driver-qemu                                  7.0.0-3                                amd64        Virtualization daemon QEMU connection driver
ii  libvirt-daemon-driver-storage-gluster                       7.0.0-3                                amd64        Virtualization daemon glusterfs storage driver
ii  libvirt-daemon-driver-storage-rbd                           7.0.0-3                                amd64        Virtualization daemon RBD storage driver
ii  libvirt-daemon-driver-storage-zfs                           7.0.0-3                                amd64        Virtualization daemon ZFS storage driver
ii  libvirt-daemon-driver-vbox                                  7.0.0-3                                amd64        Virtualization daemon VirtualBox connection driver
ii  libvirt-daemon-driver-xen                                   7.0.0-3                                amd64        Virtualization daemon Xen connection driver
ii  libvirt-daemon-system                                       7.0.0-3                                amd64        Libvirt daemon configuration files
ii  libvirt-daemon-system-systemd                               7.0.0-3                                all          Libvirt daemon configuration files (systemd)
ii  libvirt-dbus                                                1.4.0-2                                amd64        libvirt D-Bus API bindings
ii  libvirt-dev:amd64                                           7.0.0-3                                amd64        development files for the libvirt library
ii  libvirt-glib-1.0-0:amd64                                    3.0.0-1                                amd64        libvirt GLib and GObject mapping library
ii  libvirt0:amd64                                              7.0.0-3                                amd64        library for interfacing with different virtualization systems
ii  python3-libvirt                                             7.0.0-2                                amd64        libvirt Python 3 bindings
ii  ruby-vagrant-cloud                                          3.0.2-1                                all          Vagrant Cloud API Library
ii  vagrant                                                     2.2.14+dfsg-1                          all          Tool for building and distributing virtualized development environments
```

❯ vagrant plugin list
`vagrant-libvirt (0.8.2, global)`

and libvirt disk xml
```
<disk type="file" device="disk">
  <driver name="qemu" type="qcow2"/>
  <source file="/var/lib/libvirt/images/<value libvirt.default_prefix><value config.vm.define>.img" index="1"/>
  <backingStore type="file" index="2">
    <format type="qcow2"/>
    <source file="/var/lib/libvirt/images/<value box name>_vagrant_box_image_0_box.img"/>
    <backingStore/>
  </backingStore>
  <target dev="vda" bus="virtio"/>
  <alias name="ua-box-volume-0"/>
  <address type="pci" domain="0x0000" bus="0x00" slot="0x03" function="0x0"/>
</disk>
```

**Comment by electrofelix** [comment]
@slwr4thales unfortunately until someone has time to confirm how to remove backing files safely while also handling if libvirt has another VM using the same backing store, this issue will remain.

**Comment by slwr4thales** [comment]
@electrofelix is it possible to avoid/disable usage of backing file/store?

**Comment by electrofelix** [comment]
No, it's what allows the same box image to be used by multiple machines when working with a multi machine env. Changing this may be possible, but currently there is no support. Note that this path would be expensive storage wise for multi machine envs, and you could get there same behaviour by looking to delete the backing store on machine destroy instead

If you really want to remove it you could make use of the trigger functionality in vagrant to hook after the action that handles box removal to call libvirt directly to perform removal.
